File: apps/backend/app/modules/operations/schemas.py
from datetime import date, datetime
from decimal import Decimal
import uuid

from pydantic import BaseModel, ConfigDict, Field, model_validator

from app.core.enums import OperationStatus, OperationType


class OperationBase(BaseModel):
    portfolio_id: uuid.UUID
    asset_id: uuid.UUID
    operation_type: OperationType
    trade_date: date
    settlement_date: date
    quantity: Decimal = Field(gt=0)
    unit_price: Decimal = Field(ge=0)
    gross_value: Decimal | None = Field(default=None, ge=0)
    net_value: Decimal | None = Field(default=None, ge=0)
    fees: Decimal = Field(default=Decimal("0"), ge=0)
    taxes: Decimal = Field(default=Decimal("0"), ge=0)
    status: OperationStatus = OperationStatus.DRAFT
    notes: str | None = None

    @model_validator(mode="after")
    def validate_dates(self) -> "OperationBase":
        if self.settlement_date < self.trade_date:
            raise ValueError("Settlement date must be on or after trade date.")
        return self


class OperationCreate(OperationBase):
    pass


class OperationUpdate(BaseModel):
    portfolio_id: uuid.UUID | None = None
    asset_id: uuid.UUID | None = None
    operation_type: OperationType | None = None
    trade_date: date | None = None
    settlement_date: date | None = None
    quantity: Decimal | None = Field(default=None, gt=0)
    unit_price: Decimal | None = Field(default=None, ge=0)
    gross_value: Decimal | None = Field(default=None, ge=0)
    net_value: Decimal | None = Field(default=None, ge=0)
    fees: Decimal | None = Field(default=None, ge=0)
    taxes: Decimal | None = Field(default=None, ge=0)
    status: OperationStatus | None = None
    notes: str | None = None

    @model_validator(mode="after")
    def validate_dates(self) -> "OperationUpdate":
        if self.trade_date and self.settlement_date and self.settlement_date < self.trade_date:
            raise ValueError("Settlement date must be on or after trade date.")
        return self


class OperationPortfolioSummary(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: uuid.UUID
    name: str


class OperationAssetSummary(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: uuid.UUID
    ticker: str
    name: str


class OperationRead(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: uuid.UUID
    portfolio_id: uuid.UUID
    asset_id: uuid.UUID
    operation_type: OperationType
    trade_date: date
    settlement_date: date
    quantity: Decimal
    unit_price: Decimal
    gross_value: Decimal
    net_value: Decimal
    fees: Decimal
    taxes: Decimal
    status: OperationStatus
    notes: str | None
    created_at: datetime
    updated_at: datetime
    portfolio: OperationPortfolioSummary
    asset: OperationAssetSummary

File: apps/frontend/src/lib/api/schemas.ts
import { z } from "zod";

export const userSchema = z.object({
  id: z.string().uuid(),
  email: z.string(),
  full_name: z.string(),
  is_active: z.boolean(),
  is_superuser: z.boolean(),
  role: z.enum(["admin", "manager", "analyst", "viewer"]),
  created_at: z.string(),
  updated_at: z.string(),
});

export const portfolioSchema = z.object({
  id: z.string().uuid(),
  name: z.string(),
  description: z.string().nullable(),
  base_currency: z.string(),
  benchmark: z.string().nullable(),
  is_active: z.boolean(),
  created_at: z.string(),
  updated_at: z.string(),
});

export const assetSchema = z.object({
  id: z.string().uuid(),
  ticker: z.string(),
  name: z.string(),
  asset_type: z.string(),
  issuer: z.string().nullable(),
  indexer: z.string().nullable(),
  maturity_date: z.string().nullable(),
  is_active: z.boolean(),
  created_at: z.string(),
  updated_at: z.string(),
});

export const operationSchema = z.object({
  id: z.string().uuid(),
  portfolio_id: z.string().uuid(),
  asset_id: z.string().uuid(),
  operation_type: z.string(),
  trade_date: z.string(),
  settlement_date: z.string(),
  quantity: z.string(),
  unit_price: z.string(),
  gross_value: z.string(),
  net_value: z.string(),
  fees: z.string(),
  taxes: z.string(),
  status: z.string(),
  notes: z.string().nullable(),
  created_at: z.string(),
  updated_at: z.string(),
  portfolio: z.object({
    id: z.string().uuid(),
    name: z.string(),
  }),
  asset: z.object({
    id: z.string().uuid(),
    ticker: z.string(),
    name: z.string(),
  }),
});

export const cashflowEntrySchema = z.object({
  id: z.string().uuid(),
  portfolio_id: z.string().uuid(),
  operation_id: z.string().uuid().nullable(),
  entry_date: z.string(),
  settlement_date: z.string(),
  description: z.string(),
  entry_type: z.string(),
  amount: z.string(),
  status: z.string(),
  created_at: z.string(),
  updated_at: z.string(),
  portfolio: z.object({
    id: z.string().uuid(),
    name: z.string(),
  }),
  operation: z
    .object({
      id: z.string().uuid(),
      operation_type: z.string(),
      status: z.string(),
    })
    .nullable(),
});

export const assetPriceSchema = z.object({
  id: z.string().uuid(),
  asset_id: z.string().uuid(),
  price_date: z.string(),
  price: z.string(),
  source: z.string(),
  is_validated: z.boolean(),
  created_at: z.string(),
  updated_at: z.string(),
  asset: z.object({
    id: z.string().uuid(),
    ticker: z.string(),
    name: z.string(),
  }),
});

export const authTokenSchema = z.object({
  access_token: z.string(),
  token_type: z.string(),
  expires_in: z.number(),
  user: userSchema,
});

export const portfolioListSchema = z.array(portfolioSchema);
export const assetListSchema = z.array(assetSchema);
export const operationListSchema = z.array(operationSchema);
export const cashflowEntryListSchema = z.array(cashflowEntrySchema);
export const assetPriceListSchema = z.array(assetPriceSchema);
